docs: git best practices

gforcada opened the pull request:
Documentation regarding some tips and tricks on how to do better
development with git, be more respectful for the others that will
review your code and giving some advices on how to make so.

Fixes: https://github.com/plone/buildout.coredev/issues/67

jean commented:
@gforcada thanks for this.

do3cc commented on docs/git.rst:
```diff
@@ -0,0 +1,304 @@
+
+Pulling code
+------------
+*ALWAYS* (in capital letters) do a :command:`git pull -r` when fetching new code,
```
Here I wonder what -r does, but instead of explaining that, you go on with how to automate that thing

gforcada replied:
Good point, will go over it again. `-r` is to do a rebase after fetching the code from the repository.

So basically `git pull` in a vanilla git install actually does:

```
git fetch
git merge
```

With the `-r` switch you are actually doing:

```
git fetch
git rebase origin/XXX # wherever the branch you happen to be.
```

adamcheasley replied:
I agree with this point, but as @do3cc says, you should possibly mention what the argument does. Even just mentioning the word 'rebase' might be enough.

adamcheasley replied:
Sorry, ignore that last comment. I see that you have a link to further explanation. This is probably enough.

jean replied:
@adamcheasley @gforcada the option can be spelled `--rebase` instead of `-r` ... I think there's seldom any reason to use the short forms in docs.

gforcada replied:
Good point @jean 

Actually as I got the idea that it was a bit to much into the point, right now I'm writing a little introduction, hopefully it doesn't end up being a duplicate of git's own book: http://git-scm.com/book/en/v2 :)

do3cc commented on docs/git.rst:
```diff
@@ -0,0 +1,304 @@
+Always add the full URL to the issue/pull request you are fixing/referring to.
+
+Maybe within the git repository it makes sense,
+but as soon as you are outside of it, it will not.
```
I guess you rephrased something and now this sentence does not make sense right now

gforcada replied:
Most probably, as I said I wanted to get a first version out and later review it.

In this section I was referring to the fact that if your commit message says:

```
Fix issue #74
```

You can relate that issue 74 to the current project where the commit was pushed, but alas, p.a.discussion (for the sake of an example) does not have issues enabled on github, so where the hell is this issue 74 from?

My suggestion is to instead write it like:

```
Fix issue about encoding

La la la la la la la
la la la la li li li li li

Fixes issue: https://github.com/plone/Products.CMFPlone/issue/74
```

do3cc commented:
I am not a huge fan of doctoring commits after the fact. Mostly because I find it very hard to do so and I am not very proficient with it. Some things I miss: 
When I want to do clean commits I must prevent myself from just starting to modify things I notice.
I regulary start to work on something and during that work I notice I have to fix other things too. Here is an example: https://github.com/collective/collective.solr/pull/40 I would not know how to solve this.
I like the idea of changing a line only in one commit within a PR. What tools are there to help me check for this?

Finally, I am a big +1 on this, but I'd hope to receive more help on how to do this.

gforcada commented:
@do3cc  thanks for the comments and review!

About doctoring commits, I kind of understand, but still one year later, if you go stare at the code and want to understand what was going on there you will most probably end up looking at `git blame` and looking at those commits, specially for the commit messages and related files (maybe for the tests that show how the logic is supposed to work).

So at the end of the day (or year) the more effort you put on properly explaining your commits and the more curated they are, the easier it will be for other to look at them, understand what was going on and so on.

Where I work (der Freitag) we use viewlets _A LOT_, so as I'm really lazy to remember everything, I just need to remember the last time I created one and literally copy&paste&adapt the commits from the last viewlet I created because everything is there: tests, code, ZCML registration, GS profile, upgrade step, and so on.

About how to do clean commits: whenever I found something to be cleaned up I do:
- `git stash` (to save my current changes temporally)
- change back to master (if I'm working on a branch, which usually you should :))
- do the cleanups && push those changes
- get back to the branch
- rebase on top of master (`git rebase master`)
- `git stash pop` (get back the temporal changes)
- _if needed_, fix the conflicts on the code that was stashed and the cleanups that you did on master

Seems a lot of steps but is quite fast to do so.

I have another section pending to be added also about manually reverting pull requests if github doesn't allow you to do so. So I will add this cleanup strategy also.

do3cc commented:
He @gforcada thanks for the description. Worth a try and imho it would be good if this is part of the docs. 
In this particular PR I mentioned this wouldn't have been possible, but rewriting all tests is kind of big after all.

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,304 @@
+you may be doing a simple refactoring across lots of files,
+that's fine and good actually,
+as long as that commit is just and only about that refactoring,
+and not a refactoring plus moving things around plus fixing some bugs plus adding some docs plus a new cool feature plus fixing typos on documentation.
```
Ironically, this sentence is quite long and could probably be broken down into two shorter sentences. It's a little hard to follow at present.

adamcheasley commented:
Thanks for this @gforcada. I think this document is something that is desperately needed within the community.

pilz commented:
I am very happy that you did this! I strongly believe that many many people want to follow good practice but it is difficult to know what that is. Your document is a big step! Thank you!

wengole commented:
@do3cc With regards your issue of wanting changing more things than you should in one commit, you might want to read up on partial commits. UI tools help here tremendously, but it can be done from the command line quite simply too.

For UI, on Linux I use git-cola and I've seen the Mac users here (netsight) use gitx (not to be confused with gitx on Linux which is quite different)

Reference;
http://stackoverflow.com/a/1085191
http://git-scm.com/docs/git-add#_interactive_mode

gforcada committed:
git docs: basics

Brief introduction on how git works,
some concepts and its commands.

Related to: https://github.com/plone/buildout.coredev/issues/67

Many thanks to @adamcheasley for his proofreading.

gforcada committed:
git docs: guidelines

Added some guidelines on how to work more effectively with git.

These, hopefully, will help our community to move faster and better.

Related to: https://github.com/plone/buildout.coredev/issues/67

Many thanks to @adamcheasley for his proofreading.

gforcada committed:
git docs: recipes

Add some useful tips and tricks.

To be further expanded as questions come.

Related to: https://github.com/plone/buildout.coredev/issues/67

gforcada commented:
@do3cc @adamcheasley and others, I just pushed a reworked version of it, hopefully addressing your (really good) comments. If you could give it a go it would be much appreciated.

gforcada committed:
Add git docs into the index

So they can be actually found.

Related to: https://github.com/plone/buildout.coredev/issues/67

gforcada commented:
I see that a proof read is more than needed, but at 2am I don't feel like it, so please, bear with the grammar errors on the meantime, sorry!

adamcheasley commented:
Thanks @gforcada. This looks really good. Like you say, there are a few grammatical errors, but I held off commenting on those until you've had time for a proof read.

gforcada commented:
@adamcheasley it would be really appreciated (no native English speaker here as you can see) if you can either fix or just add comments on them in github, thanks in advance!

adamcheasley commented:
No problem, will do.

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+
+The index
+  A temporal staging storage with changes on files that are pending to be added to a commit.
+  If your git output is colored,
```
"coloured"! "colored" is the US spelling.

do3cc replied:
Do we have enough us speakers to make readable for them too? SCNR ;-)

davisagli replied:
As a developer on international open source projects I speak fluent butchered English. :-p

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+Branches
+--------
+Another great feature of DVCS is cheap branching,
+i.e. make it effortlessly and really useful to use branches.
```
I think this would read better as:
"i.e. branching in git is effortless and really useful."

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+--------
+Another great feature of DVCS is cheap branching,
+i.e. make it effortlessly and really useful to use branches.
+No more working on the master branch always as is too much effort to work on branches.
```
Possibly consider changing to:
"As it's no longer too much effort to branch, there is no need to always work on the master branch. A developer can branch easily for each fix/feature."

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+i.e. make it effortlessly and really useful to use branches.
+No more working on the master branch always as is too much effort to work on branches.
+
+Branches allows you to tinker with your changes while keeping the master branch clean.
```
"Branches allow you". There is an erroneous 's' on the end of allow there.

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+Branches allows you to tinker with your changes while keeping the master branch clean.
+
+Not only that,
+it also allows to keep modifying your changes until you and your peers are fine with them.
```
"it also allows you to keep..."

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+
+   If there are files on the index,
+   or files not tracked by git,
+   or the status of your branch in regards of the remote one,
```
"or the status of your local branch with regards to the remote"

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+      If git is criticized for being complex,
+      this command is one of the main sources of complains.
+
+      It is so *powerful* yet apparently so *random* that makes people crazy.
```
I'm not entirely clear on this sentence. There's nothing really random about this command. It might be helpful to compare this command to svn's 'switch'.

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+
+This way you do not introduce new extra commits and the git history is kept as simple as possible.
+
+This is specially important when trying to understand why some changes were made,
```
"This is especially important..."

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+that an upgrade should be needed...
+
+Remember that the first code review is the one you do on your own.
+Some inspiration/better phrased:
```
should be "better phrasing"

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+
+Turns instead into a list of questions:
+
+- why this interface is renamed here if has nothing to do with this adapter?
```
"Why is this interface renamed here if it has nothing to do with this adapter?"

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+- fixing typos on documentation
+- pep8 fixes
+
+Is completely fine to make refactorings.
```
"It is absolutely fine to refactor."

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+
+Following this advice will:
+
+- keep things simple where there's no gain on adding complexity
```
_in_ adding complexity...

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+
+- keep things simple where there's no gain on adding complexity
+- make your changes easy to be reviewed
+- make later on lookups on those changes easy to follow
```
Could you clarify this last point for me?

gforcada replied:
Sorry for the late answer: I'm trying to say that if "one commit does one thing" if you ever have to do something similar (say create a viewlet) it will make your life easier as it will be a matter of copy/pasting that previous commit that you can look it up.

Does that make sense?

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+wrote that in 2009,
+so most probably once a refactor of that package is done later on 2015,
+Martin is no longer around,
+and if he would be,
```
"and if he was,"

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+get one project that you worked 5 years ago,
+get a random commit and:
+
+- see if by just reading the commit message already gives enough information of what changes are in
```
"See if, just by reading the commit message, you are given enough information of what changes have been made, when comparing the commit message and the actual code. Does the commit message match the code changed?"

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+
+See the history example above: :ref:`general-guidelines-label`.
+
+Unfortunately the flat view from GitHub prevents to see that,
```
"Unfortunately the flat view from GitHub prevents us from seeing that"

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+-------------------------
+*On the context of a pull request*
+
+Ask yourself: which relation has a cleanup commit,
```
"Ask yourself: What relation does a cleanup commit, say pep8 fixes or other code analysis fixes, have with your pull request?"

adamcheasley commented on docs/git.rst:
```diff
@@ -0,0 +1,611 @@
+Only tells you that the author did not take care at all about the one who will review it,
+and specially about the person that in +5 years will try to understand that test.
+Specially because now the test is not only spread between 4 commits,
+but most probably during those 5 years it already had some refactorings,
```
"but most probably during those 5 years it has already been refactored,"

adamcheasley commented:
Done! I've just realised, I could have helped you out a bit there and just made the changes in the branch. If you'd like me to do that, just let me know.

gforcada commented:
@adamcheasley I just pushed your fixes, many thanks!

If no one objects I think it can be merged.

do3cc commented:
heh, I was too slow.
I finished the reading and I am happy with the recipes extension.
What I am not sure is if the doc followed the guidelines, especially this one:
http://docs.plone.org/about/styleguide.html#line-length-translations

At first I thought you tried to stay within 80 chars but then I saw that this is not the case. Right now some phrases are written in so many tiny lines that it is hard to read in source. Like here: https://github.com/plone/buildout.coredev/pull/71/files#diff-5a9477c2b3ec283ab80fd30c2fa4726dR69
